### PR TITLE
feat: introduce row trait

### LINKED ISSUE #6432: Introduce the row trait and utilities.

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR introduces the `Row2` trait and replace most `Vec<Datum>` with `impl Row2` in state table and storage table interfaces.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- Close #6432 
- #6433

### CODE REVIEW

File: src/common/src/row/mod.rs [hunk 1]
```diff
@@ -12,21 +12,341 @@
...
+
+    fn to_owned_row(&self) -> Row {
+        Row(self.iter().map(|d| d.to_owned_datum()).collect())
+    }
+
+    fn into_owned_row(self) -> Row {
+        self.to_owned_row()
+    }
+
+    fn value_serialize(&self) -> Vec<u8> {
```
Comment (yuhao-su) [inline src/common/src/row/mod.rs:51]: Is there any specific reason to call it `value_serialize`? Maybe `value_encoding` or serialize is more clear and succinct.

Comment (BugenZhao) [reply src/common/src/row/mod.rs]: I name it `serialize` to follow the convention of Rust libraries. And `value` is to show that it should be used in the value of kv store. We may add a `key_serialize` or `ordered_serialize` in the future as well.

**Review by TennyZhuang — APPROVED** [review]
LGTM

**Review by yuhao-su — APPROVED** [review]

**Review by wcy-fdu — APPROVED** [review]
LGTM! Good Work!

**Review by st1page — APPROVED** [review]

**Pull request merged**